src/pq/param.cr

```diff
@@ -1,164 +1,292 @@
+require "../pg/interval"
 require "../pg/geo"
 
 module PQ
   # :nodoc:
-  record Param, slice : Slice(UInt8), size : Int32, format : Int16 do
+  record Param, slice : Slice(UInt8), size : Int32, format : Format do
+    enum Format : Int16
+      None   = -1
+      Text   =  0
+      Binary =  1
+    end
     delegate to_unsafe, to: slice
 
     #  Internal wrapper to represent an encoded parameter
 
     def self.encode(val : Nil)
-      binary Pointer(UInt8).null.to_slice(0), -1
+      binary Bytes.empty, -1
+    end
+
+    def self.encode(val : Bool, into slice : Bytes = Bytes.new(1))
+      slice[0] = val ? 1u8 : 0u8
+      binary slice
+    end
+
+    def self.encode(val : String)
+      encode val.to_slice
+    end
+
+    def self.encode(val : String, into slice : Bytes)
+      encode val.to_slice, into: slice
     end
 
     def self.encode(val : Slice)
-      binary val, val.size
+      binary val
     end
 
-    def self.encode(val : Array)
-      text encode_array(val)
+    def self.encode(val : Slice, into slice : Bytes)
+      val.copy_to slice
+
+      binary slice
+    end
+
+    def self.encode(val : Array(T)) forall T
+      bytes = ArrayEncoder.new(val).to_slice
+
+      binary bytes
     end
 
     def self.encode(val : Time)
-      text Time::Format::RFC_3339.format(val, fraction_digits: 9)
+      # text Time::Format::RFC_3339.format(val, fraction_digits: 9)
+      encode ((val - 30.years).to_unix_ns // 1_000).to_i64
     end
 
+    {% for type in %w[Int16 Int32 Int64 Float32 Float64] %}
+      def self.encode(val : {{type.id}}, into slice : Bytes = Bytes.new(sizeof(typeof(val))))
+        IO::ByteFormat::NetworkEndian.encode val, slice
```

Comment:
Whoever added the ability encode numeric values directly into slices without an intermediate `IO` gets the world's biggest high-five from me.